### DISCUSSION TITLE
Different results when reading an SVG from file or from blob

### DISCUSSION THREAD

**Discussion by xoich** [ImageMagick/ImageMagick#3547] [Help]
I'm using PHP Imagick with Image Magick version 6.9.11-60, The SVG in question has some SVG2 features. When I add it from file the output is as expected, with SVG2 features working. When instead I add the SVG from a string blob the result is different and SVG2 features are not respected.

I would expect this two snippets of code to have the same results, but they don't:
```
$img->readImage('test.svg');
```
```
$svg = file_get_contents('test.svg');
$img->readImageBlob($svg, 'svg');    // same result when not passing the second parameter
```

How can I add the SVG from a blob and still have SVG2 features working?

**Comment by snibgo** [comment]
What SVG v2 features are you using?

Perhaps your two methods use different SVG delegates. This is one of: internal MSVG, internal RSVG, or external Inkscape. (Or is could even be external RSVG). Of these, I expect only Inkscape has SVG v2 features.

As far as I know, SVG v2 isn't a standard.

"Verbose" text output will tell you which rasterizer is used. I don't know how to do "verbose" in PHP.

EDIT: What is "SVG2"? Perhaps you mean "SVG v1.2", which isn't a standard.

**Reply by xoich** [reply]
You are right I'm pretty sure it is using Inkscape as delegate in one case and not in the other. 

The SVG2 feature is `shape-inside` style for text. You can have automatic line-wrapping with it. Is there a way to have line-wrapping without requiring Inkscape as the actual converter?

**Comment by snibgo** [comment]
As far as I know, there is no "SVG2".

As far as I know, no version of SVG has a "shape-inside" feature.

In SVG draft v1.2, text can be flowed within paths. I show examples at [SVG text](http://im.snibgo.com/svgtext.htm#areas).

Perhaps you are confusing SVG and CSS. A draft version of CSS has "shape-inside". See https://drafts.csswg.org/css-shapes-2/#shape-inside-property

**Reply by xoich** [reply]
I'm refering to this: https://www.w3.org/TR/2015/WD-SVG2-20150915/text.html#TextShapeInside
SVG2 is only a draft, but apparently Inkscape uses some of its features.

> In SVG draft v1.2, text can be flowed within paths. I show examples at SVG text.

Thanks for this, I'll check it. Unfortunately it seems it does not work with the other SVG delegates other than Inkscape.

**Comment by snibgo** [comment]
Thanks for the correction. "SVG 2" does exist, as a draft, and it includes "shape-inside" as an attribute of the "text" tag."

We can copy-paste the CSS shape example on the page you linked, into a file called shins.svg:
```
<svg xmlns="http://www.w3.org/2000/svg"
     width="300" height="100" viewBox="0 0 300 100">

  <text shape-inside="circle(100 at 100 100)"
	font-family="sans-serif"
	font-size="20px">Lorem ipsum dolor sit amet, consec-tetuer adipiscing elit...</text>

</svg>
```
... and try to rasterize that with Inkscape. On my computer, the version is "1.0 (4035a4fb49, 2020-05-01)". Inkscape reports errors:
```
SPIShapes::read: Invalid shape value: circle(100
SPIShapes::read: Invalid shape value: at
SPIShapes::read: Invalid shape value: 100
SPIShapes::read: Invalid shape value: 100)
```
So that didn't work. Another example uses an SVG shape, whch we save as shins2.svg:
```
<svg xmlns="http://www.w3.org/2000/svg"
     width="300" height="100" viewBox="0 0 300 100">

  <defs>
    <rect id="wrap" x="50" y="10" width="200" height="80"/>
  </defs>

  <text shape-inside="url(#wrap)"
	font-family="sans-serif"
	font-size="20px" >This text wraps in a rectangle.</text>

</svg>
```
This does work:
```
inkscape -o x.png shins2.svg
```
![x](https://user-images.githubusercontent.com/33812368/114876031-d15ccb80-9df5-11eb-887b-2181d1348733.png)


Sadly, doing this from ImageMagick, with Inkscape as delegate, makes the wrong result. I don't know why.

EDIT: Note that a later version of "SVG 2" is at https://www.w3.org/TR/SVG2/
